lib/storage/providers/aws.rb

```diff
@@ -0,0 +1,148 @@
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# Flight Storage is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with Flight Storage. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on Flight Storage, please visit:
+# https://github.com/openflighthpc/flight-storage
+#==============================================================================
+require 'aws-sdk-s3'
+
+require_relative '../client'
+require_relative '../tree'
+
+module Storage
+  class AWSClient < Client
+    FRIENDLY_NAME = "Amazon S3"
+  
+    def self.creds_schema
+      {
+        access_key: String,
+        secret_access_key: String,
+        region: String,
+        bucket_name: String
+      }
+    end
+    
+    def list(path="/", tree: false)
+      path = path.delete_prefix("/")
+      subtree = dir_tree.dig(*path.split("/"))
+      msg = ""
+      if tree
+        msg << subtree.show
+      else
+        subtree.subdirs.each do |child|
+          msg << child.name + "/\n"
+        end
+        msg << subtree.files.join("\n")
+      end
+      msg
+    end
+    
+    def dir_tree
+      @dir_tree ||= Tree.new("/", self.to_hash("")[nil])
+    end
+    
+    def pull(source, dest)
+      if dir_tree.file_exists?(source)
+        source = source.delete_prefix("/")
+        resp = client.get_object(response_target: dest, bucket: @credentials[:bucket_name], key: source)
+      else
+        raise ResourceNotFoundError, source
+      end
+      dest
+    end
+    
+    def push(source, dest)
+      dest = dest.delete_prefix("/")
+      if dir_tree.file_exists?(dest)
+        raise ResourceExistsError, dest
+      end
+      
+      obj = Aws::S3::Object.new(bucket_name: @credentials[:bucket_name], key: dest, :client => client)
+      obj.upload_stream({part_size: 100 * 1024 * 1024, tempfile: true}) do |write_stream|
+        IO.copy_stream(File.open(source, "rb"), write_stream)
+      end
+    end
+    
+    def delete(path)
+      path = path.delete_prefix("/")
+      if dir_tree.file_exists?(path)
+        client.delete_object(bucket: @credentials[:bucket_name], key: path)
+      else
+        raise ResourceNotFoundError, path
+      end
+    end
+    
+    # Convert the bucket contents into a tree-like hash
+    def to_hash(prefix)
+      children = []
+      resp = client.list_objects(bucket: @credentials[:bucket_name], delimiter: "/", prefix: prefix)
```

Comment:
Is there a reason we're using `Aws::S3::Client#list_objects` instead of `Aws::S3::Client#list_objects_v2`?